### PR TITLE
rangefeed: Resilience + iterator limit

### LINKED ISSUE #32947: changefeedccl: Implement ChangeAggregator buffer + rangefeed manager
Track implementation of a central ChangeAggregator output buffer and a rangefeed manager.  The intention of this is to prevent slow or temporarily unavailable output sinks from causing unnecessary retry-related thrashing in the rangefeed system.

Detailed problem and proposal outlined in #32944.

### PR DESCRIPTION
**Pull request by mrtracy**
Commit enhances distsender.RangeFeed() with two important features:

1. Adds significant resilience in the case of slow consumers; slow
consuming rangefeeds can be restarted directly at this level and can run
as slowly as needed.
2. Adds an "iterator limit" sempahore which limits the number of
rangefeeds that can be in the "Catchup" phase at a single time, which
has the effect of limiting the impact on the cluster of creating too
many engine iterators. This is analogous to the similar semaphore-based
limiting used previously by the poll-based changefeed system.

Resolves #32947

Release note: None

### CODE REVIEW

**Review by danhhz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34108)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz, @mrtracy, and @nvanbenschoten)

___
*[pkg/kv/dist_sender_rangefeed.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-LWWd5qD3JsUM1qtaxv-:-LWWd5qD3JsUM1qtaxv0:b-92tb9d) ([raw file](https://github.com/cockroachdb/cockroach/blob/bca207a91ab1bf51ae1a69df826cc987f47b7873/pkg/kv/dist_sender_rangefeed.go#L41)):*
> ```Go
> // RangeFeed to each of the individual ranges. It streams back results on the
> // provided channel.
> func (ds *DistSender) RangeFeed(
> ```

The changefeed user of this ends up having to call it for a bunch of non-adjacent spans, because we schedule the processor with whatever spans it's the leaseholder for. How about if you pass a non-nil `limitIteratorSem`, it uses it?

We also may want some limiting on the server side, in case there are many clients all starting up at once, but that can wait for another pr.
___
*[pkg/kv/dist_sender_rangefeed.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-LWWd_RuAsMzU2hCtdXo:-LWWd_RuAsMzU2hCtdXp:b-bixw5s) ([raw file](https://github.com/cockroachdb/cockroach/blob/bca207a91ab1bf51ae1a69df826cc987f47b7873/pkg/kv/dist_sender_rangefeed.go#L72)):*
> ```Go
> 	g := ctxgroup.WithContext(ctx)
> 	// Goroutine that processes subdivided ranges and creates a changefeed for
> 	// each, while using the semaphore to limiting the number of rangefeeds that
> ```

nit: to limit
___
*[pkg/kv/dist_sender_rangefeed.go, line 78 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-LWWf3rc4ETKahRxvLYP:-LWWf3rc4ETKahRxvLYQ:b-qhr8dr) ([raw file](https://github.com/cockroachdb/cockroach/blob/bca207a91ab1bf51ae1a69df826cc987f47b7873/pkg/kv/dist_sender_rangefeed.go#L78)):*
> ```Go
> 			select {
> 			case sri := <-rangeCh:
> 				// Semaphore management. If a semaphore is being used to limit
> ```

nit: this select spans quite a few lines and has a select inside it, which makes it hard to follow the structure. can you move the body of this `case` down below the select?
___
*[pkg/kv/dist_sender_rangefeed.go, line 219 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-LWWfcQHArgwIsY__561:-LWWfcQHArgwIsY__562:b6zb2v1) ([raw file](https://github.com/cockroachdb/cockroach/blob/bca207a91ab1bf51ae1a69df826cc987f47b7873/pkg/kv/dist_sender_rangefeed.go#L219)):*
> ```Go
> 					return err, false
> 				}
> 				releaseSem()
> ```

Wouldn't a bunch of these double-release the sem? What if it finishes the catchup, then later the range splits?
___
*[pkg/roachpb/errors.proto, line 472 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-LWWduHu55rGb9xnqc8q:-LWWduHu55rGb9xnqc8r:bxx6fck) ([raw file](https://github.com/cockroachdb/cockroach/blob/bca207a91ab1bf51ae1a69df826cc987f47b7873/pkg/roachpb/errors.proto#L472)):*
> ```protobuf
>     // A Raft command was missing a logical operation log.
>     REASON_LOGICAL_OPS_MISSING = 4;
>     // The consumer was processing events to slowly to keep up with live raft
> ```

nit: too slowly

**Review by mrtracy — COMMENTED** [review]
Dan's suggestion to move the limiter to the server side was too good to pass up. I've done another pass at this where all of the semaphore limiting is done on the server-side, with a limit set per store which seems very natural.

RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-:-LWYYt9rDTzMyfn5kj1J:b-5khb74)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz, @mrtracy, and @nvanbenschoten)

___
*[pkg/kv/dist_sender_rangefeed.go, line 41 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-LWWd5qD3JsUM1qtaxv-:-LWYZAWD4z-OaiXg2Ed0:buaklrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/bca207a91ab1bf51ae1a69df826cc987f47b7873/pkg/kv/dist_sender_rangefeed.go#L41)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

The changefeed user of this ends up having to call it for a bunch of non-adjacent spans, because we schedule the processor with whatever spans it's the leaseholder for. How about if you pass a non-nil `limitIteratorSem`, it uses it?

We also may want some limiting on the server side, in case there are many clients all starting up at once, but that can wait for another pr.
</blockquote></details>

Refactored to use this suggestion instead of  using a semaphore in the distsender, thanks!

**Review by danhhz — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-:-LX-bVBcCAVhhosZDDnX:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @danhhz, @mrtracy, and @nvanbenschoten)

___
*[pkg/storage/replica_rangefeed.go, line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-LX-_T790QL-HrpCnWQx:-LX-_T790QL-HrpCnWQy:b-3door5) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b08c3bf89c49104432fc6317b2a644fe3303f3b/pkg/storage/replica_rangefeed.go#L166)):*
> ```Go
> 			return roachpb.NewError(err)
> 		}
> 	}
> ```

I'm worried that this structure doesn't make it obvious enough that the Begins and Finishes match up. How about a defer based pattern:

```
usingCatchupIter := false
var iterFinish func()
if !args.Timestamp.IsEmpty() {
    usingCatchupIter = true
    if err := iteratorLimiter.Begin(ctx); err != nil {
        return roachpb.NewError(err)
    }
    iterFinish = iteratorLimiter.Finish
    defer func() {
        if iterFinish != nil {
            iterFinish()
        }
    }()
}
```

then below, you'd transfer the "ownership" of finishing to `semaphoreLimitedIterator` (which now gets a `func()` instead of a `limit.ConcurrentRequestLimiter`:
```
if usingCatchupIter {
    innerIter := r.Engine().NewIterator(engine.IterOptions{
        UpperBound:       args.Span.EndKey,
        MinTimestampHint: args.Timestamp,
    })
    catchUpIter = semaphoreLimitedIterator{
        SimpleIterator: innerIter,
        semFinish:            iterFinish,
    }
    iterFinish = nil
}
```

**Review by mrtracy — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34108)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @danhhz and @nvanbenschoten)

___
*[pkg/storage/replica_rangefeed.go, line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-LX-_T790QL-HrpCnWQx:-LXCSDYa9XOe1cwoekJ_:b-8gilzh) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b08c3bf89c49104432fc6317b2a644fe3303f3b/pkg/storage/replica_rangefeed.go#L166)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

I'm worried that this structure doesn't make it obvious enough that the Begins and Finishes match up. How about a defer based pattern:

```
usingCatchupIter := false
var iterFinish func()
if !args.Timestamp.IsEmpty() {
    usingCatchupIter = true
    if err := iteratorLimiter.Begin(ctx); err != nil {
        return roachpb.NewError(err)
    }
    iterFinish = iteratorLimiter.Finish
    defer func() {
        if iterFinish != nil {
            iterFinish()
        }
    }()
}
```

then below, you'd transfer the "ownership" of finishing to `semaphoreLimitedIterator` (which now gets a `func()` instead of a `limit.ConcurrentRequestLimiter`:
```
if usingCatchupIter {
    innerIter := r.Engine().NewIterator(engine.IterOptions{
        UpperBound:       args.Span.EndKey,
        MinTimestampHint: args.Timestamp,
    })
    catchUpIter = semaphoreLimitedIterator{
        SimpleIterator: innerIter,
        semFinish:            iterFinish,
    }
    iterFinish = nil
}
```
</blockquote></details>

Done, used the pattern you suggested.

**Review by danhhz — COMMENTED** [review]
Thanks, matt!

bors r=danhhz

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34108#-:-LXGmugM9sFw-9q6HC2j:b-91ca05)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @danhhz and @nvanbenschoten)
